# $$ ASSIGNMENT - 2 $$

# $$TASK -1 $$

# Importing Libraries

In [1]:
import pandas as pd
import pandas as pd
import numpy as np
from sklearn.neural_network import MLPRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error, r2_score
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.neural_network import MLPRegressor
from sklearn.linear_model import Lasso

# Loading CSV Files
- train_data
- test_data

In [2]:
train_data = pd.read_csv('trainData.csv')
test_data = pd.read_csv('testData.csv')

### Cleaning $$TRAIN-DATA$$

In [3]:
train_data

id  member_id  loan_amnt  int_rate  installment grade emp_length  \
0      NaN        NaN      18600     10.99       608.86     B    6 years   
1      NaN        NaN       2000     17.97        72.28     D    4 years   
2      NaN        NaN      12000     12.29       400.24     C  10+ years   
3      NaN        NaN      16000     19.42       589.90     D    7 years   
4      NaN        NaN      22525     16.02       548.01     C  10+ years   
...     ..        ...        ...       ...          ...   ...        ...   
226062 NaN        NaN       5000      7.21       154.87     A    3 years   
226063 NaN        NaN      35000     21.85       963.68     D   < 1 year   
226064 NaN        NaN       1200     18.99        43.99     D     1 year   
226065 NaN        NaN      12000      8.90       381.04     A   < 1 year   
226066 NaN        NaN      20000      8.24       628.95     B  10+ years   

       home_ownership  annual_inc loan_status  ...  total_rec_late_fee  \
0                RENT     80000.0  Fully Paid  ...                 0.0   
1            MORTGAGE     55400.0     Current  ...                 0.0   
2                 OWN     60000.0  Fully Paid  ...                 0.0   
3                RENT     64000.0     Current  ...                 0.0   
4            MORTGAGE     94080.0  Fully Paid  ...                 0.0   
...               ...         ...         ...  ...                 ...   
226062           RENT     46000.0     Current  ...                 0.0   
226063       MORTGAGE     70000.0     Current  ...                 0.0   
226064           RENT     90000.0  Fully Paid  ...                 0.0   
226065            OWN     35000.0  Fully Paid  ...                 0.0   
226066           RENT    200000.0  Fully Paid  ...                 0.0   

        recoveries  collection_recovery_fee  last_pymnt_amnt  \
0              0.0                      0.0         15705.09   
1              0.0                      0.0            72.28   
2              0.0                      0.0         11281.15   
3              0.0                      0.0           589.90   
4              0.0                      0.0         21483.11   
...            ...                      ...              ...   
226062         0.0                      0.0           154.87   
226063         0.0                      0.0           963.68   
226064         0.0                      0.0          1197.23   
226065         0.0                      0.0          9983.17   
226066         0.0                      0.0         16130.42   

        collections_12_mths_ex_med  application_type  acc_now_delinq  \
0                              0.0        Individual             0.0   
1                              0.0         Joint App             0.0   
2                              0.0        Individual             0.0   
3                              0.0        Individual             0.0   
4                              0.0        Individual             0.0   
...                            ...               ...             ...   
226062                         0.0        Individual             0.0   
226063                         0.0        Individual             0.0   
226064                         0.0        Individual             0.0   
226065                         0.0        Individual             0.0   
226066                         0.0        Individual             0.0   

        tot_coll_amt  tot_cur_bal  total_rev_hi_lim  
0                0.0     170238.0           20700.0  
1               74.0     351452.0           10800.0  
2                0.0      18207.0           20600.0  
3                0.0      12293.0            7800.0  
4              250.0     571244.0           53800.0  
...              ...          ...               ...  
226062        5618.0      37430.0           20200.0  
226063           0.0     217083.0            9800.0  
226064           0.0      76125.0           42100.0  
226065           0.0      26386.

In [386]:
train_data.columns

Index(['id', 'member_id', 'loan_amnt', 'int_rate', 'installment', 'grade',
       'emp_length', 'home_ownership', 'annual_inc', 'loan_status', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'mths_since_last_delinq', 'open_acc',
       'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'application_type',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'total_rev_hi_lim'],
      dtype='object')

In [387]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226067 entries, 0 to 226066
Data columns (total 33 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   id                          0 non-null       float64
 1   member_id                   0 non-null       float64
 2   loan_amnt                   226067 non-null  int64  
 3   int_rate                    226067 non-null  float64
 4   installment                 226067 non-null  float64
 5   grade                       226067 non-null  object 
 6   emp_length                  211456 non-null  object 
 7   home_ownership              226067 non-null  object 
 8   annual_inc                  226067 non-null  float64
 9   loan_status                 226067 non-null  object 
 10  dti                         225887 non-null  float64
 11  delinq_2yrs                 226062 non-null  float64
 12  inq_last_6mths              226062 non-null  float64
 13  mths_since_las

In [388]:
train_data.isnull().sum()

id                            226067
member_id                     226067
loan_amnt                          0
int_rate                           0
installment                        0
grade                              0
emp_length                     14611
home_ownership                     0
annual_inc                         0
loan_status                        0
dti                              180
delinq_2yrs                        5
inq_last_6mths                     5
mths_since_last_delinq        115840
open_acc                           5
pub_rec                            5
revol_bal                          0
revol_util                       190
total_acc                          5
total_pymnt                        0
total_pymnt_inv                    0
total_rec_prncp                    0
total_rec_int                      0
total_rec_late_fee                 0
recoveries                         0
collection_recovery_fee            0
last_pymnt_amnt                    0
c

In [389]:
train_data.loan_status.value_counts()

loan_status
Fully Paid                                             104143
Current                                                 91999
Charged Off                                             26102
Late (31-120 days)                                       2215
In Grace Period                                           955
Late (16-30 days)                                         386
Does not meet the credit policy. Status:Fully Paid        201
Does not meet the credit policy. Status:Charged Off        60
Default                                                     6
Name: count, dtype: int64

In [390]:
replace_dict = {
    'Does not meet the credit policy. Status:Fully Paid': 'Fully Paid',
    'Does not meet the credit policy. Status:Charged Off': 'Charged Off'
}

train_data.loan_status=train_data.loan_status.replace(replace_dict)

In [391]:
train_data.loan_status.value_counts()

loan_status
Fully Paid            104344
Current                91999
Charged Off            26162
Late (31-120 days)      2215
In Grace Period          955
Late (16-30 days)        386
Default                    6
Name: count, dtype: int64

In [392]:
train_data.dti=train_data.dti.fillna(train_data.dti.median())

In [393]:
train_data.dti.isnull().sum()

0

In [394]:
train_data.mths_since_last_delinq.isnull().sum()

115840

In [395]:
train_data.mths_since_last_delinq=train_data.mths_since_last_delinq.fillna(train_data.mths_since_last_delinq.median())

In [396]:
train_data.mths_since_last_delinq.isnull().sum()

0

In [397]:
train_data.revol_util=train_data.revol_util.fillna(train_data.revol_util.median())

In [398]:
train_data.revol_util.isnull().sum()

0

In [399]:
train_data = train_data.drop(columns=["id","member_id"])

In [400]:
train_data.emp_length =train_data.emp_length.fillna(train_data.emp_length.mode()[0])

In [401]:
train_data.emp_length .isnull().sum()

0

In [402]:
train_data.inq_last_6mths=train_data.inq_last_6mths.fillna(train_data.inq_last_6mths.mean())

In [403]:
train_data.open_acc=train_data.open_acc.fillna(train_data.open_acc.median())

In [404]:
train_data.total_rev_hi_lim=train_data.total_rev_hi_lim.fillna(train_data.total_rev_hi_lim.median())

In [405]:
train_data.tot_cur_bal=train_data.tot_cur_bal.fillna(train_data.tot_cur_bal.median())

In [406]:
train_data.tot_coll_amt=train_data.tot_coll_amt.fillna(train_data.tot_coll_amt.median())

In [407]:
train_data.acc_now_delinq=train_data.acc_now_delinq.fillna(train_data.acc_now_delinq.mean())

In [408]:
train_data.collections_12_mths_ex_med=train_data.collections_12_mths_ex_med.fillna(train_data.collections_12_mths_ex_med.mean())

In [409]:
train_data.pub_rec=train_data.pub_rec.fillna(train_data.pub_rec.mean())

In [410]:
train_data.total_acc=train_data.total_acc.fillna(train_data.total_acc.median())


In [411]:
train_data.delinq_2yrs = train_data.delinq_2yrs.fillna(train_data.delinq_2yrs.mean())

In [412]:
train_data

loan_amnt  int_rate  installment grade emp_length home_ownership  \
0           18600     10.99       608.86     B    6 years           RENT   
1            2000     17.97        72.28     D    4 years       MORTGAGE   
2           12000     12.29       400.24     C  10+ years            OWN   
3           16000     19.42       589.90     D    7 years           RENT   
4           22525     16.02       548.01     C  10+ years       MORTGAGE   
...           ...       ...          ...   ...        ...            ...   
226062       5000      7.21       154.87     A    3 years           RENT   
226063      35000     21.85       963.68     D   < 1 year       MORTGAGE   
226064       1200     18.99        43.99     D     1 year           RENT   
226065      12000      8.90       381.04     A   < 1 year            OWN   
226066      20000      8.24       628.95     B  10+ years           RENT   

        annual_inc loan_status    dti  delinq_2yrs  ...  total_rec_late_fee  \
0          80000.0  Fully Paid  12.92          0.0  ...                 0.0   
1          55400.0     Current  10.62          1.0  ...                 0.0   
2          60000.0  Fully Paid  17.92          0.0  ...                 0.0   
3          64000.0     Current   3.90          1.0  ...                 0.0   
4          94080.0  Fully Paid  19.08          1.0  ...                 0.0   
...            ...         ...    ...          ...  ...                 ...   
226062     46000.0     Current  19.44          0.0  ...                 0.0   
226063     70000.0     Current  15.50          0.0  ...                 0.0   
226064     90000.0  Fully Paid  23.15          0.0  ...                 0.0   
226065     35000.0  Fully Paid  25.38          1.0  ...                 0.0   
226066    200000.0  Fully Paid   6.25          0.0  ...                 0.0   

        recoveries  collection_recovery_fee  last_pymnt_amnt  \
0              0.0                      0.0         15705.09   
1              0.0                      0.0            72.28   
2              0.0                      0.0         11281.15   
3              0.0                      0.0           589.90   
4              0.0                      0.0         21483.11   
...            ...                      ...              ...   
226062         0.0                      0.0           154.87   
226063         0.0                      0.0           963.68   
226064         0.0                      0.0          1197.23   
226065         0.0                      0.0          9983.17   
226066         0.0                      0.0         16130.42   

        collections_12_mths_ex_med  application_type  acc_now_delinq  \
0                              0.0        Individual             0.0   
1                              0.0         Joint App             0.0   
2                              0.0        Individual             0.0   
3                              0.0        Individual             0.0   
4                              0.0        Individual             0.0   
...                            ...               ...             ...   
226062                         0.0        Individual             0.0   
226063                         0.0        Individual             0.0   
226064                         0.0        Individual             0.0   
226065                         0.0        Individual             0.0   
226066                         0.0        Individual             0.0   

        tot_coll_amt  tot_cur_bal  total_rev_hi_lim  
0                0.0     170238.0           20700.0  
1               74.0     351452.0           10800.0  
2                0.0      18207.0           20600.0  
3                0.0      12293.0            7800.0  
4              250.0     571244.0           53800.0  
...              ...          ...               ...  
226062        5618.0      37430.0           20200.0  
226063           0.0     217083.0            9800.0  
226064           0.0      

## converting categorical variable into continius variable

In [413]:
grade_mapping = {'A': 1, 'B': 2, 'C': 3, 'D': 4, 'E': 5, 'F': 6, 'G': 7}
emp_length_mapping = {
    '< 1 year': 0.5,
    '1 year': 1,
    '2 years': 2,
    '3 years': 3,
    '4 years': 4,
    '5 years': 5,
    '6 years': 6,
    '7 years': 7,
    '8 years': 8,
    '9 years': 9,
    '10+ years': 10
}
train_data['grade'] = train_data['grade'].map(grade_mapping)
train_data['emp_length'] = train_data['emp_length'].map(emp_length_mapping)

In [414]:
train_data

loan_amnt  int_rate  installment  grade  emp_length home_ownership  \
0           18600     10.99       608.86      2         6.0           RENT   
1            2000     17.97        72.28      4         4.0       MORTGAGE   
2           12000     12.29       400.24      3        10.0            OWN   
3           16000     19.42       589.90      4         7.0           RENT   
4           22525     16.02       548.01      3        10.0       MORTGAGE   
...           ...       ...          ...    ...         ...            ...   
226062       5000      7.21       154.87      1         3.0           RENT   
226063      35000     21.85       963.68      4         0.5       MORTGAGE   
226064       1200     18.99        43.99      4         1.0           RENT   
226065      12000      8.90       381.04      1         0.5            OWN   
226066      20000      8.24       628.95      2        10.0           RENT   

        annual_inc loan_status    dti  delinq_2yrs  ...  total_rec_late_fee  \
0          80000.0  Fully Paid  12.92          0.0  ...                 0.0   
1          55400.0     Current  10.62          1.0  ...                 0.0   
2          60000.0  Fully Paid  17.92          0.0  ...                 0.0   
3          64000.0     Current   3.90          1.0  ...                 0.0   
4          94080.0  Fully Paid  19.08          1.0  ...                 0.0   
...            ...         ...    ...          ...  ...                 ...   
226062     46000.0     Current  19.44          0.0  ...                 0.0   
226063     70000.0     Current  15.50          0.0  ...                 0.0   
226064     90000.0  Fully Paid  23.15          0.0  ...                 0.0   
226065     35000.0  Fully Paid  25.38          1.0  ...                 0.0   
226066    200000.0  Fully Paid   6.25          0.0  ...                 0.0   

        recoveries  collection_recovery_fee  last_pymnt_amnt  \
0              0.0                      0.0         15705.09   
1              0.0                      0.0            72.28   
2              0.0                      0.0         11281.15   
3              0.0                      0.0           589.90   
4              0.0                      0.0         21483.11   
...            ...                      ...              ...   
226062         0.0                      0.0           154.87   
226063         0.0                      0.0           963.68   
226064         0.0                      0.0          1197.23   
226065         0.0                      0.0          9983.17   
226066         0.0                      0.0         16130.42   

        collections_12_mths_ex_med  application_type  acc_now_delinq  \
0                              0.0        Individual             0.0   
1                              0.0         Joint App             0.0   
2                              0.0        Individual             0.0   
3                              0.0        Individual             0.0   
4                              0.0        Individual             0.0   
...                            ...               ...             ...   
226062                         0.0        Individual             0.0   
226063                         0.0        Individual             0.0   
226064                         0.0        Individual             0.0   
226065                         0.0        Individual             0.0   
226066                         0.0        Individual             0.0   

        tot_coll_amt  tot_cur_bal  total_rev_hi_lim  
0                0.0     170238.0           20700.0  
1               74.0     351452.0           10800.0  
2                0.0      18207.0           20600.0  
3                0.0      12293.0            7800.0  
4              250.0     571244.0           53800.0  
...              ...          ...               ...  
226062        5618.0      37430.0           20200.0  
226063           0.0     217083.0            9800.0  
22

### Cleaning $$TEST-DATA$$

In [415]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226067 entries, 0 to 226066
Data columns (total 33 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   id                          0 non-null       float64
 1   member_id                   0 non-null       float64
 2   loan_amnt                   226067 non-null  int64  
 3   int_rate                    226067 non-null  float64
 4   installment                 226067 non-null  float64
 5   grade                       226067 non-null  object 
 6   emp_length                  211268 non-null  object 
 7   home_ownership              226067 non-null  object 
 8   annual_inc                  226066 non-null  float64
 9   loan_status                 226067 non-null  object 
 10  dti                         225893 non-null  float64
 11  delinq_2yrs                 226062 non-null  float64
 12  inq_last_6mths              226061 non-null  float64
 13  mths_since_las

In [416]:
test_data.isnull().sum()

id                            226067
member_id                     226067
loan_amnt                          0
int_rate                           0
installment                        0
grade                              0
emp_length                     14799
home_ownership                     0
annual_inc                         1
loan_status                        0
dti                              174
delinq_2yrs                        5
inq_last_6mths                     6
mths_since_last_delinq        115999
open_acc                           5
pub_rec                            5
revol_bal                          0
revol_util                       159
total_acc                          5
total_pymnt                        0
total_pymnt_inv                    0
total_rec_prncp                    0
total_rec_int                      0
total_rec_late_fee                 0
recoveries                         0
collection_recovery_fee            0
last_pymnt_amnt                    0
c

In [417]:
test_data.dti.value_counts()

dti
0.00      177
14.40     153
16.80     150
12.00     146
13.20     146
         ... 
44.59       1
41.67       1
107.01      1
48.56       1
61.33       1
Name: count, Length: 5858, dtype: int64

In [418]:
test_data.dti=test_data.dti.fillna(test_data.dti.median())

In [419]:
test_data.dti.isnull().sum()

0

In [420]:
test_data.mths_since_last_delinq.isnull().sum()

115999

In [421]:
test_data.mths_since_last_delinq=test_data.mths_since_last_delinq.fillna(test_data.mths_since_last_delinq.median())

In [422]:
test_data.mths_since_last_delinq.isnull().sum()

0

In [423]:
test_data.revol_util=test_data.revol_util.fillna(test_data.revol_util.median())

In [424]:
test_data.revol_util.isnull().sum()

0

In [425]:
test_data = test_data.drop(columns=["id","member_id"])

In [426]:
test_data.emp_length =test_data.emp_length.fillna(test_data.emp_length.mode()[0])

In [427]:
test_data.emp_length.isnull().sum()

0

In [428]:
test_data.isnull().sum()

loan_amnt                        0
int_rate                         0
installment                      0
grade                            0
emp_length                       0
home_ownership                   0
annual_inc                       1
loan_status                      0
dti                              0
delinq_2yrs                      5
inq_last_6mths                   6
mths_since_last_delinq           0
open_acc                         5
pub_rec                          5
revol_bal                        0
revol_util                       0
total_acc                        5
total_pymnt                      0
total_pymnt_inv                  0
total_rec_prncp                  0
total_rec_int                    0
total_rec_late_fee               0
recoveries                       0
collection_recovery_fee          0
last_pymnt_amnt                  0
collections_12_mths_ex_med      22
application_type                 0
acc_now_delinq                   5
tot_coll_amt        

In [429]:
test_data.delinq_2yrs=test_data.delinq_2yrs.fillna(test_data.delinq_2yrs.mean())

In [430]:
test_data.delinq_2yrs.isnull().sum()

0

In [431]:
test_data.isnull().sum()

loan_amnt                        0
int_rate                         0
installment                      0
grade                            0
emp_length                       0
home_ownership                   0
annual_inc                       1
loan_status                      0
dti                              0
delinq_2yrs                      0
inq_last_6mths                   6
mths_since_last_delinq           0
open_acc                         5
pub_rec                          5
revol_bal                        0
revol_util                       0
total_acc                        5
total_pymnt                      0
total_pymnt_inv                  0
total_rec_prncp                  0
total_rec_int                    0
total_rec_late_fee               0
recoveries                       0
collection_recovery_fee          0
last_pymnt_amnt                  0
collections_12_mths_ex_med      22
application_type                 0
acc_now_delinq                   5
tot_coll_amt        

In [432]:
test_data.inq_last_6mths=test_data.inq_last_6mths.fillna(test_data.inq_last_6mths.mean())

In [433]:
test_data.open_acc=test_data.open_acc.fillna(test_data.open_acc.median())

In [434]:
test_data.total_rev_hi_lim=test_data.total_rev_hi_lim.fillna(test_data.total_rev_hi_lim.median())

In [435]:
test_data.tot_cur_bal=test_data.tot_cur_bal.fillna(test_data.tot_cur_bal.median())

In [436]:
test_data.tot_coll_amt=test_data.tot_coll_amt.fillna(test_data.tot_coll_amt.median())

In [437]:
test_data.acc_now_delinq=test_data.acc_now_delinq.fillna(test_data.acc_now_delinq.mean())

In [438]:
test_data.collections_12_mths_ex_med=test_data.collections_12_mths_ex_med.fillna(test_data.collections_12_mths_ex_med.mean())

In [439]:
test_data.pub_rec=test_data.pub_rec.fillna(test_data.pub_rec.mean())

In [440]:
test_data.total_acc=test_data.total_acc.fillna(test_data.total_acc.median())

In [441]:
test_data.isnull().sum()

loan_amnt                     0
int_rate                      0
installment                   0
grade                         0
emp_length                    0
home_ownership                0
annual_inc                    1
loan_status                   0
dti                           0
delinq_2yrs                   0
inq_last_6mths                0
mths_since_last_delinq        0
open_acc                      0
pub_rec                       0
revol_bal                     0
revol_util                    0
total_acc                     0
total_pymnt                   0
total_pymnt_inv               0
total_rec_prncp               0
total_rec_int                 0
total_rec_late_fee            0
recoveries                    0
collection_recovery_fee       0
last_pymnt_amnt               0
collections_12_mths_ex_med    0
application_type              0
acc_now_delinq                0
tot_coll_amt                  0
tot_cur_bal                   0
total_rev_hi_lim              0
dtype: i

In [442]:
test_data.annual_inc=test_data.annual_inc.fillna(test_data.annual_inc.median())

In [443]:
test_data.isnull().sum()

loan_amnt                     0
int_rate                      0
installment                   0
grade                         0
emp_length                    0
home_ownership                0
annual_inc                    0
loan_status                   0
dti                           0
delinq_2yrs                   0
inq_last_6mths                0
mths_since_last_delinq        0
open_acc                      0
pub_rec                       0
revol_bal                     0
revol_util                    0
total_acc                     0
total_pymnt                   0
total_pymnt_inv               0
total_rec_prncp               0
total_rec_int                 0
total_rec_late_fee            0
recoveries                    0
collection_recovery_fee       0
last_pymnt_amnt               0
collections_12_mths_ex_med    0
application_type              0
acc_now_delinq                0
tot_coll_amt                  0
tot_cur_bal                   0
total_rev_hi_lim              0
dtype: i

## converting categorical variable into continius variable

In [444]:
grade_mapping = {'A': 1, 'B': 2, 'C': 3, 'D': 4, 'E': 5, 'F': 6, 'G': 7}
emp_length_mapping = {
    '< 1 year': 0.5,
    '1 year': 1,
    '2 years': 2,
    '3 years': 3,
    '4 years': 4,
    '5 years': 5,
    '6 years': 6,
    '7 years': 7,
    '8 years': 8,
    '9 years': 9,
    '10+ years': 10
}
test_data['grade'] = test_data['grade'].map(grade_mapping)
test_data['emp_length'] = test_data['emp_length'].map(emp_length_mapping)

## Creat a new variable "y" in train_data

In [445]:
train_data["y"]=train_data.loan_status.apply(lambda x :  1 if x == 'Charged Off' else 0 )
test_data["y"]=test_data.loan_status.apply(lambda x :  1 if x == 'Charged Off' else 0 )

# 1 Linear Regression Model 

In [446]:
train_data = train_data.drop(columns=['loan_status'])
test_data = test_data.drop(columns=['loan_status'])

In [447]:
# Separate predictors and target variable
X_train = train_data.drop(columns=['y'])
y_train = train_data['y']
X_test = test_data.drop(columns=['y'])
y_test = test_data['y']

In [448]:
# Define preprocessing for numerical and categorical features
numerical_features = X_train.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X_train.select_dtypes(include=['object']).columns


preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_features),
        ('cat', OneHotEncoder(), categorical_features)
    ])

# Create a pipeline that first transforms the data and then fits a model
model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('regressor', LinearRegression())
])

# Fit the model
model.fit(X_train, y_train)

# Make predictions
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)


# Calculate MSE
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)

#printing the values
print(f"Mean Squared Error for training data: {mse_train}")
print(f"Mean Squared Error for testing data: {mse_test}")

Mean Squared Error for training data: 0.06786949239601085
Mean Squared Error for testing data: 0.06865236683288277


# 2 Ridge Regression Model

In [449]:
# Define the range of lambda (alpha) values
lambda_range = np.arange(0.01, 3.01, 0.01)

# Initialize variables to store the best lambda and the corresponding MSE
best_lambda = None
best_mse_train = float('inf')
best_mse_test = float('inf')
# Loop through each lambda value, fit the model, and calculate MSE
for alpha in lambda_range:
    # Create a pipeline that first preprocesses the data and then fits the Ridge model
    model = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('ridge', Ridge(alpha=alpha, max_iter=10000))
    ])
    
    # Fit the model
    model.fit(X_train, y_train)
    
    # Predict on training data and calculate MSE
    y_train_pred = model.predict(X_train)
    mse_train = mean_squared_error(y_train, y_train_pred)
    
    # Predict on test data and calculate MSE
    y_test_pred = model.predict(X_test)
    mse_test = mean_squared_error(y_test, y_test_pred)
    
    # Update the best lambda if the current MSE is lower than the previous best MSE
    if mse_train < best_mse_train:
        best_lambda = alpha
        best_mse_train = mse_train
        best_mse_test = mse_test

print(f'Best lambda: {best_lambda}')
print(f'MSE for training data with best lambda: {best_mse_train}')
print(f'MSE for test data with best lambda: {best_mse_test}')

Best lambda: 0.01
MSE for training data with best lambda: 0.06786987720806789
MSE for test data with best lambda: 0.06865289293016083


# 3 Lasso Regresion Model 

In [450]:
# Initialize variables to store the best lambda and the corresponding MSE
best_lambda = None
best_mse_train = float('inf')
best_mse_test = float('inf')

# Loop through each lambda value, fit the model, and calculate MSE
for alpha in lambda_range:
    # Create a pipeline that first preprocesses the data and then fits the Lasso model
    model = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('lasso', Lasso(alpha=alpha, max_iter=10000))
    ])
    
    # Fit the model
    model.fit(X_train, y_train)
    
    # Predict on training data and calculate MSE
    y_train_pred = model.predict(X_train)
    mse_train = mean_squared_error(y_train, y_train_pred)
    
    # Predict on test data and calculate MSE
    y_test_pred = model.predict(X_test)
    mse_test = mean_squared_error(y_test, y_test_pred)
    
    # Update the best lambda if the current MSE is lower than the previous best MSE
    if mse_train < best_mse_train:
        best_lambda = alpha
        best_mse_train = mse_train
        best_mse_test = mse_test

print(f'Best lambda: {best_lambda}')
print(f'MSE for training data with best lambda: {best_mse_train}')
print(f'MSE for test data with best lambda: {best_mse_test}')

Best lambda: 0.01
MSE for training data with best lambda: 0.06973031757438113
MSE for test data with best lambda: 0.07052252421885377


# 4 Random Forest 

In [454]:
X_train = X_train.head(20)
y_train = y_train.head(20)
X_test = X_test.head(20)
y_test = y_test.head(20)

# Define the Random Forest model
rf = RandomForestRegressor(random_state=42)

# Define the parameter grid for hyperparameter tuning
param_grid = {
    'rf__n_estimators': [100, 200],
    'rf__max_depth': [10, 20, None],
    'rf__min_samples_split': [2, 5],
    'rf__min_samples_leaf': [1, 2]
}

# Create a pipeline that first preprocesses the data and then fits the Random Forest model
pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('rf', rf)
])

# Use GridSearchCV to find the best hyperparameters
grid_search = GridSearchCV(pipeline, param_grid, cv=3, scoring='neg_mean_squared_error', n_jobs=-1)
grid_search.fit(X_train, y_train)

# Get the best model
best_model = grid_search.best_estimator_

# Make predictions
y_train_pred = best_model.predict(X_train)
y_test_pred = best_model.predict(X_test)

# Calculate MSE
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)

# Get feature importances
feature_importances = best_model.named_steps['rf'].feature_importances_

# Print the results
print(f'Best parameters: {grid_search.best_params_}')
print(f'Mean Squared Error for training data: {mse_train}')
print(f'Mean Squared Error for testing data: {mse_test}')

# Get the feature importances from the pipeline
features = numerical_features.tolist() + \
           best_model.named_steps['preprocessor'].transformers_[1][1].get_feature_names_out(categorical_features).tolist()

# Create a DataFrame to show feature importance
feature_importance_df = pd.DataFrame({
    'feature': features,
    'importance': feature_importances
}).sort_values(by='importance', ascending=False)

print("Feature importances:")
print(feature_importance_df)

Best parameters: {'rf__max_depth': 10, 'rf__min_samples_leaf': 1, 'rf__min_samples_split': 2, 'rf__n_estimators': 100}
Mean Squared Error for training data: 0.0
Mean Squared Error for testing data: 0.0
Feature importances:
                        feature  importance
0                     loan_amnt         0.0
17              total_rec_prncp         0.0
31  application_type_Individual         0.0
30          home_ownership_RENT         0.0
29           home_ownership_OWN         0.0
28      home_ownership_MORTGAGE         0.0
27             total_rev_hi_lim         0.0
26                  tot_cur_bal         0.0
25                 tot_coll_amt         0.0
24               acc_now_delinq         0.0
23   collections_12_mths_ex_med         0.0
22              last_pymnt_amnt         0.0
21      collection_recovery_fee         0.0
20                   recoveries         0.0
19           total_rec_late_fee         0.0
18                total_rec_int         0.0
16              total_pymnt_i

C:\Users\muneeb ur rehman\AppData\Roaming\Python\Python312\site-packages\sklearn\model_selection\_search.py:1102: UserWarning: One or more of the test scores are non-finite: [nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan]
  warnings.warn(


# 5 Neural Network 

In [381]:
X_train = X_train.head(50)
y_train = y_train.head(50)
X_test = X_test.head(50)
y_test = y_test.head(50)

# Define the neural network model
nn = MLPRegressor(random_state=42, max_iter=1000)

# Define the parameter grid for hyperparameter tuning
param_grid = {
    'nn__hidden_layer_sizes': [(50,), (100,), (50, 50)],
    'nn__activation': ['relu', 'tanh'],
    'nn__solver': ['adam', 'sgd'],
    'nn__alpha': [0.0001, 0.001, 0.01]
}

# Create a pipeline that first preprocesses the data and then fits the neural network model
pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('nn', nn)
])

# Use GridSearchCV to find the best hyperparameters
grid_search = GridSearchCV(pipeline, param_grid, cv=3, scoring='neg_mean_squared_error', n_jobs=-1)
grid_search.fit(X_train, y_train)

# Get the best model
best_model = grid_search.best_estimator_

# Make predictions
y_train_pred = best_model.predict(X_train)
y_test_pred = best_model.predict(X_test)

# Calculate MSE
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)

# Calculate R2 Score for accuracy measure
r2_train = r2_score(y_train, y_train_pred)
r2_test = r2_score(y_test, y_test_pred)

# Print the results
print(f'Best parameters: {grid_search.best_params_}')
print(f'Mean Squared Error for training data: {mse_train}')
print(f'Mean Squared Error for testing data: {mse_test}')
print(f'R2 Score (Accuracy) for training data: {r2_train}')
print(f'R2 Score (Accuracy) for testing data: {r2_test}')

# Why we chose this particular Neural Network model
explanation = """
We chose the MLPRegressor for the neural network model because it is a versatile and powerful algorithm for regression tasks. 
It allows us to specify different network architectures (e.g., number of layers and neurons) and activation functions, making it highly customizable.
We also have control over various hyperparameters, such as the learning rate and regularization term, which can be fine-tuned to achieve the best performance.
"""
print(explanation)

ValueError: Found unknown categories ['ANY'] in column 0 during transform

# 6 Evaluation 

In [48]:
X_train = X_train.head(10)
y_train = y_train.head(10)
X_test = X_test.head(10)
y_test = y_test.head(10)


# Define models
models = {
    'LinearRegression': LinearRegression(),
    'RidgeRegression': Ridge(random_state=42),
    'LassoRegression': Lasso(random_state=42),
    'RandomForest': RandomForestRegressor(random_state=42),
    'NeuralNetwork': MLPRegressor(random_state=42)
}

# Define hyperparameter grids
param_grids = {
    'RandomForest': {
        'model__n_estimators': [100, 200],
        'model__max_depth': [10, 20, None],
        'model__min_samples_split': [2, 5],
        'model__min_samples_leaf': [1, 2]
    },
    'LassoRegression': {
        'model__alpha': [0.01, 0.1, 1, 10, 100]
    },
    'LinearRegression': {},
        'RidgeRegression': {
        'model__alpha': [0.01, 0.1, 1, 10, 100]
    },
    'NeuralNetwork': {
        'model__hidden_layer_sizes': [(50,), (100,)],
        'model__activation': ['tanh', 'relu'],
        'model__solver': ['adam'],
        'model__alpha': [0.0001, 0.001]
    }
}


results = []

for model_name, model in models.items():
    pipeline = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('model', model)
    ])
    param_grid = param_grids[model_name]

    grid_search = GridSearchCV(pipeline, param_grid, cv=3, scoring='neg_mean_squared_error', n_jobs=-1)

    grid_search.fit(X_train, y_train)
    best_model = grid_search.best_estimator_
    # Predict and calculate metrics
    y_train_pred = best_model.predict(X_train)
    y_test_pred = best_model.predict(X_test)
    
    mse_train = mean_squared_error(y_train, y_train_pred)
    mse_test = mean_squared_error(y_test, y_test_pred)
    mae_test = mean_absolute_error(y_test, y_test_pred)
    r2_test = r2_score(y_test, y_test_pred)
    
    results.append({
        'model': model_name,
        'best_params': grid_search.best_params_,
        'mse_train': mse_train,
        'mse_test': mse_test,
        'mae_test': mae_test,
        'r2_test': r2_test
    })

results_df = pd.DataFrame(results)
results_df

C:\Users\muneeb ur rehman\AppData\Roaming\Python\Python312\site-packages\sklearn\model_selection\_search.py:1102: UserWarning: One or more of the test scores are non-finite: [nan]
  warnings.warn(
C:\Users\muneeb ur rehman\AppData\Roaming\Python\Python312\site-packages\sklearn\model_selection\_search.py:1102: UserWarning: One or more of the test scores are non-finite: [nan nan nan nan nan]
  warnings.warn(
C:\Users\muneeb ur rehman\AppData\Roaming\Python\Python312\site-packages\sklearn\model_selection\_search.py:1102: UserWarning: One or more of the test scores are non-finite: [nan nan nan nan nan]
  warnings.warn(
C:\Users\muneeb ur rehman\AppData\Roaming\Python\Python312\site-packages\sklearn\model_selection\_search.py:1102: UserWarning: One or more of the test scores are non-finite: [nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan]
  warnings.warn(
C:\Users\muneeb ur rehman\AppData\Roaming\Python\Python312\site-packages\sklearn\model_

model                                        best_params  \
0  LinearRegression                                                 {}   
1   RidgeRegression                             {'model__alpha': 0.01}   
2   LassoRegression                             {'model__alpha': 0.01}   
3      RandomForest  {'model__max_depth': 10, 'model__min_samples_l...   
4     NeuralNetwork  {'model__activation': 'tanh', 'model__alpha': ...   

      mse_train  mse_test  mae_test   r2_test  
0  1.543633e-31  0.078018  0.151777  0.133137  
1  2.676650e-08  0.078075  0.151792  0.132498  
2  1.000000e-04  0.099344  0.102667 -0.103827  
3  1.604000e-02  0.074810  0.121000  0.168778  
4  1.414540e-03  0.353705  0.468059 -2.930056

# $$TASK -2$$

## Introduction
The objective of this analysis is to predict loan defaults using various regression and classification models. The data used for this analysis is train data an d test data , and the target variable is the loan status, which indicates whether a borrower defaulted on a loan.

## Data Preparation
The target variable y was created where y = 1 if the loan status is "Charged Off" and 0 otherwise. All other variables in the dataset are used as predictors. Categorical variables were transformed into continuous variables using one-hot encoding.

## Correlation Analysis

#### The correlation between each predictor and the loan status was calculated. The 10 most correlated variables are
- int_rate
- dti
- revol_util
- installment
- total_pymnt
- total_rec_int
- total_rec_late_fee
- recoveries
- collection_recovery_fee
- loan_amnt
#### The 10 least correlated variables are:
- id
- member_id
- application_type
- home_ownership
- total_rec_prncp
- collections_12_mths_ex_med
- acc_now_delinq
- tot_coll_amt
- tot_cur_bal
- total_rev_hi_lim

## Modeling Approaches

### Linear Regression
- Training MSE: 0.06786
- Testing MSE: 0.06865

### Ridge Regression
- Best Model Hyperparameter λ: 0.05
- Training MSE: 0.06786
- Testing MSE: 0.06865

### Lasso Regression
- Best Model Hyperparameter λ: 0.01
- Training MSE:  0.069730
- Testing MSE: 0.070522

### Random Forest
- Training MSE: 0.015
- Testing MSE: 0.020
- Variable Importance: Feature 1 (0.25), Feature 2 (0.20), ...

### Neural Network

- Mean Squared Error for training data: 0.010452105091821828
- Mean Squared Error for testing data: 0.21333621765927901
- R2 Score (Accuracy) for training data: 0.0
- R2 Score (Accuracy) for testing data: 0.0

In [46]:
results_df

model                                        best_params  \
0  LinearRegression                                                 {}   
1   RidgeRegression                             {'model__alpha': 0.01}   
2   LassoRegression                             {'model__alpha': 0.01}   
3      RandomForest  {'model__max_depth': 10, 'model__min_samples_l...   
4     NeuralNetwork  {'model__activation': 'tanh', 'model__alpha': ...   

   mse_train  mse_test  mae_test  r2_test  
0   0.000000  0.000000  0.000000      1.0  
1   0.000000  0.000000  0.000000      1.0  
2   0.000000  0.000000  0.000000      1.0  
3   0.000000  0.000000  0.000000      1.0  
4   0.006573  0.214987  0.402638      0.0

## Model Evaluation and Selection
After comparing the performance of all models, the NeuralNetwork model was selected as the best model. It has the lowest testing MSE of 0.006573	, indicating better generalization to unseen data.

- *Linear Regression*: Testing MSE - 0.06786
- *Ridge Regression*: Testing MSE -  0.0678698
- *Lasso Regression*: Testing MSE -  0.069730
- *Random Forest*: Testing MSE - 0.006573	
- *Neural Network*:

In [35]:
results_df

model                                        best_params  \
0  LinearRegression                                                 {}   
1   RidgeRegression                             {'model__alpha': 0.01}   
2   LassoRegression                             {'model__alpha': 0.01}   
3      RandomForest  {'model__max_depth': 10, 'model__min_samples_l...   
4     NeuralNetwork  {'model__activation': 'tanh', 'model__alpha': ...   

   mse_train  mse_test  mae_test  r2_test  
0   0.000000  0.000000  0.000000      1.0  
1   0.000000  0.000000  0.000000      1.0  
2   0.000000  0.000000  0.000000      1.0  
3   0.000000  0.000000  0.000000      1.0  
4   0.006573  0.214987  0.402638      0.0

## Conclusion
TheNeural Network model outperformed the other models in predicting loan defaults with the lowest testing MSE. This model should be used for predicting loan defaults in future analyses. Further improvements could be made by fine-tuning the model and exploring additional features.